### PR TITLE
Detailed train outputs

### PR DESCRIPTION
**Pull request by lgatys**
Changes proposed in this pull request:

- The `TorchTrainStep` now enables monitoring arbitrary model outputs during training.
- `TorchTrainEngine.forward_train` now returns a tuple `loss, model_outputs` for each micro batch.
- The list of model outputs for all micro batches in a batch is passed to the `TrainCallback.log_batch` and `TrainCallback.post_batch`.
- The documentation in the `log_batch` and `post_batch` functions  of the `TrainCallback` for why `batch_outputs` is `List[Dict[str, Any]]` is copied from the `pre_batch` function which has the same format for the `batch` input.

## Before submitting

- [x] I've read and followed all steps in the [Making a pull request](https://github.com/allenai/tango/blob/main/CONTRIBUTING.md#making-a-pull-request)
    section of the `CONTRIBUTING` docs.
- [x] I've updated or added any relevant docstrings following the syntax described in the
    [Writing docstrings](https://github.com/allenai/tango/blob/main/CONTRIBUTING.md#writing-docstrings) section of the `CONTRIBUTING` docs.

## After submitting

- [ ] All GitHub Actions jobs for my pull request have passed.

### CODE REVIEW

File: tango/integrations/torch/train.py [hunk 1]
```diff
@@ -607,25 +607,30 @@ def save_state(step: int):
...
                 # Get loss.
-                micro_batch_loss = training_engine.forward_train(
+                micro_batch_loss, micro_batch_outputs = training_engine.forward_train(
                     micro_batch, micro_batch_idx, len(batch)
                 )
                 if torch.isnan(micro_batch_loss):
                     raise ValueError("nan loss encountered")
                 batch_loss += micro_batch_loss.detach().item()
+                batch_outputs.append(
+                    {key: output.detach() for key, output in micro_batch_outputs.items()}
```
Comment (epwalsh) [inline tango/integrations/torch/train.py:620]: Good idea with detaching ✅

We should handle non-tensor outputs as well:

```suggestion
                    {key: output.detach() if isinstance(output, torch.Tensor) else output for key, output in micro_batch_outputs.items()}
```

**Review by epwalsh — CHANGES_REQUESTED** [review]
Looks great! Just one minor change requested

**Comment by lgatys** [comment]
> Looks great! Just one minor change requested

Thanks! I added your suggestion and fixed the tests. I think the remaining ones that are failing have to be re-run from your end?

**Review by epwalsh — APPROVED** [review]
LGTM!
